File: backend/app/schemas/organization.py
from datetime import datetime
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, Field


class MembershipRole(StrEnum):
    OWNER = "OWNER"
    ADMIN = "ADMIN"
    MEMBER = "MEMBER"


class OrganizationCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    slug: str = Field(min_length=1, max_length=100, pattern=r"^[a-z0-9-]+$")
    description: str | None = None


class OrganizationUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None


class OrganizationResponse(BaseModel):
    id: UUID
    name: str
    slug: str
    description: str | None
    owner_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class OrganizationMemberAdd(BaseModel):
    user_id: UUID
    role: MembershipRole = MembershipRole.MEMBER


class OrganizationMemberResponse(BaseModel):
    id: UUID
    organization_id: UUID
    user_id: UUID
    role: MembershipRole
    created_at: datetime

    model_config = {"from_attributes": True}

File: apps/mobile/lib/features/organizations/data/models/organization.dart
import 'package:flutter/foundation.dart';

@immutable
class Organization {
  const Organization({
    required this.id,
    required this.name,
    required this.slug,
    required this.description,
    required this.ownerId,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String name;
  final String slug;
  final String? description;
  final String ownerId;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory Organization.fromJson(Map<String, dynamic> json) => Organization(
        id: json['id'] as String,
        name: json['name'] as String,
        slug: json['slug'] as String,
        description: json['description'] as String?,
        ownerId: json['owner_id'] as String,
        createdAt: DateTime.parse(json['created_at'] as String),
        updatedAt: DateTime.parse(json['updated_at'] as String),
      );
}

enum MembershipRole {
  owner,
  admin,
  member;

  String toJson() => switch (this) {
        MembershipRole.owner => 'OWNER',
        MembershipRole.admin => 'ADMIN',
        MembershipRole.member => 'MEMBER',
      };

  factory MembershipRole.fromJson(String value) => switch (value.toUpperCase()) {
        'OWNER' => MembershipRole.owner,
        'ADMIN' => MembershipRole.admin,
        'MEMBER' => MembershipRole.member,
        _ => throw FormatException('Unknown membership role: $value'),
      };
}

@immutable
class OrganizationMember {
  const OrganizationMember({
    required this.id,
    required this.organizationId,
    required this.userId,
    required this.role,
    required this.createdAt,
  });

  final String id;
  final String organizationId;
  final String userId;
  final MembershipRole role;
  final DateTime createdAt;

  factory OrganizationMember.fromJson(Map<String, dynamic> json) =>
      OrganizationMember(
        id: json['id'] as String,
        organizationId: json['organization_id'] as String,
        userId: json['user_id'] as String,
        role: MembershipRole.fromJson(json['role'] as String),
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}
